moda-contracts/src/Marketplace.sol

```diff
@@ -0,0 +1,236 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.21;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+import "@openzeppelin/contracts/token/ERC1155/IERC1155.sol";
+import "@openzeppelin/contracts/token/ERC1155/utils/ERC1155Holder.sol";
+import "../src/interfaces/IMarketplace.sol";
+import "../src/interfaces/Releases/IReleasesRegistration.sol";
+
+/**
+ * @title Marketplace
+ * @dev This contract allows buying and selling of Releases and charges a fee on each sale.
+ */
+contract Marketplace is IMarketplace, ERC1155Holder, ReentrancyGuard, AccessControl {
+    using SafeERC20 for IERC20;
+
+    // State Variables
+
+    address _usdc;
+    address _catalog;
+    address payable public treasury;
+    uint256 public treasuryFee;
+
+    /// @dev seller => saleId => Sale
+    mapping(address => Sale[]) private sales;
+
+    // Errors
+    error CannotBeZeroAddress();
+    error TreasuryFeeCannotBeZero();
+    error TokenAmountCannotBeZero();
+    error MaxCountCannotBeZero();
+    error StartCannotBeAfterEnd(uint256 startTime, uint256 endTime);
+    error InsufficientSupply(uint256 remainingSupply);
+    error MaxSupplyReached(uint256 maxSupplyPerWallet);
+    error SaleNotStarted(uint256 startTime, uint256 currentTime);
+    error SaleHasEnded(uint256 endTime, uint256 currentTime);
+    error ReleasesIsNotRegistered();
+
+    /**
+     * @dev Constructor
+     * @param treasury_ - The address of the organizations treasury
+     * @param treasuryFee_ - The percentage that will be transferred
+     * to the treasury on each sale. Based on a denominator of 10_000 e.g. 1000 = 10%
+     * @param usdc - The USDC token address
+     * @param catalog - The address of the catalog contract
+     */
+    constructor(address payable treasury_, uint256 treasuryFee_, address usdc, address catalog) {
```

Comment:
usdc could be `IERC20 token`.

catalog should be typed to an interface required by the Marketplace contract. e.g. `ICatalog`, or a smaller interface that sufficiently describes all the functions required by the contract.